internal/controller/sparkconnect/reconciler.go

```diff
@@ -0,0 +1,565 @@
+func (r *Reconciler) createOrUpdateConfigMap(ctx context.Context, conn *v1alpha1.SparkConnect) error {
+	logger := ctrl.LoggerFrom(ctx)
+	logger.V(1).Info("Create or update ConfigMap")
+
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GetConfigMapName(conn),
+			Namespace: conn.Namespace,
+		},
+	}
+
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, cm, func() error {
+		if err := r.mutateConfigMap(ctx, conn, cm); err != nil {
+			return fmt.Errorf("failed to mutate configmap: %v", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create or update configmap: %v", err)
+	}
+
+	return nil
+}
+
+// mutateConfigMap mutates the configmap for the SparkConnect resource.
+func (r *Reconciler) mutateConfigMap(ctx context.Context, conn *v1alpha1.SparkConnect, cm *corev1.ConfigMap) error {
+	if cm.Labels == nil {
+		cm.Labels = map[string]string{}
+
+	}
+	cm.Labels[common.LabelCreatedBySparkOperator] = "true"
+	cm.Labels[common.LabelSparkConnName] = conn.Name
+
+	if err := ctrl.SetControllerReference(conn, cm, r.scheme); err != nil {
+		return fmt.Errorf("failed to set controller reference")
+	}
+
+	podTemplateData, err := yaml.Marshal(conn.Spec.Executor.Template)
+	if err != nil {
+		return fmt.Errorf("failed to marshal executor pod template: %v", err)
+	}
+	if cm.Data == nil {
+		cm.Data = map[string]string{}
+	}
+	cm.Data[ExecutorPodTemplateFileName] = string(podTemplateData)
+	return nil
+}
+
+// createOrUpdateServerPod creates or updates the server pod for the SparkConnect resource.
+func (r *Reconciler) createOrUpdateServerPod(ctx context.Context, conn *v1alpha1.SparkConnect) error {
+	logger := ctrl.LoggerFrom(ctx)
+	logger.V(1).Info("Create or update server pod")
+
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GetServerPodName(conn),
+			Namespace: conn.Namespace,
+		},
+	}
+
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, pod, func() error {
+		// Mutate server pod.
+		if err := r.mutateServerPod(ctx, conn, pod); err != nil {
+			return fmt.Errorf("failed to build server pod: %v", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create or update server pod: %v", err)
+	}
+
+	// Update SparkConnect status.
+	ready := util.IsPodReady(pod)
+	if ready {
+		condition := metav1.Condition{
+			Type:    string(v1alpha1.SparkConnectConditionServerPodReady),
+			Status:  metav1.ConditionTrue,
+			Reason:  string(v1alpha1.SparkConnectConditionServerPodReady),
+			Message: "Server pod is ready",
+		}
+		_ = meta.SetStatusCondition(&conn.Status.Conditions, condition)
+		conn.Status.State = v1alpha1.SparkConnectStateReady
+	} else {
+		condition := metav1.Condition{
+			Type:    string(v1alpha1.SparkConnectConditionServerPodReady),
+			Status:  metav1.ConditionFalse,
+			Reason:  string(v1alpha1.SparkConnectConditionServerPodReady),
+			Message: "Server pod is not ready",
```

Comment:
Any benefit to passing message from pod here? e.g. `fmt.Sprintf("Server pod is not ready %s", pod.Status.Message)`